### PR TITLE
8242032: G1 region remembered sets may contain non-coarse level PRTs for already coarsened regions

### PR DESCRIPTION
**Pull request by jaokim**
This fix adds a check for coarsened region in mutex guarded section, when adding a reference to a remembered set.

Haven't been able to produce a testcase -- please advice on how to, or if not necessary.

**Testing:**
* hs-tier, hs-tier2

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8242032](https://bugs.openjdk.java.net/browse/JDK-8242032): G1 region remembered sets may contain non-coarse level PRTs for already coarsened regions

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.java.net/census#ayang) (@albertnetymk - Author) ⚠️ Review applies to 7398c89b2c1e6a660451088615dd0ddfe0257903
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2545/head:pull/2545`
`$ git checkout pull/2545`

### CODE REVIEW

File: src/hotspot/share/gc/g1/heapRegionRemSet.cpp [hunk 1]
```diff
@@ -145,6 +145,13 @@ void OtherRegionsTable::add_reference(OopOrNarrowOopStar from, uint tid) {
   PerRegionTable* prt = find_region_table(ind, from_hr);
   if (prt == NULL) {
     MutexLocker x(_m, Mutex::_no_safepoint_check_flag);
+
+    // Make sure region hasn't been coarsened by other thread.
```
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/heapRegionRemSet.cpp:149]: Maybe mentioning this is an intentional re-check under lock, sth like, "Rechecking if the region is coarsened while holding the lock."

**Review by albertnetymk — APPROVED** [review]

File: src/hotspot/share/gc/g1/heapRegionRemSet.cpp (hunk 1 above)
Comment (jaokim) [reply src/hotspot/share/gc/g1/heapRegionRemSet.cpp]: Yes, thank you!

**Comment by jaokim** [comment]
Thank you @albertnetymk and @tschatzl for review!

/integrate

**Pull request closed without merging**